core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1111,51 +1112,85 @@ final class Stream[+F[_], +O] private (private val free: FreeC[Algebra[Nothing,
       implicit timer: Timer[F2],
       F: Concurrent[F2]): Stream[F2, Chunk[O]] =
     Stream
-      .eval(Queue.synchronousNoneTerminated[F2, Either[Token, Chunk[O]]])
-      .flatMap { q =>
-        def startTimeout: Stream[F2, Token] =
-          Stream.eval(F.delay(new Token)).flatTap { t =>
-            Stream.supervise { timer.sleep(d) *> q.enqueue1(t.asLeft.some) }
-          }
+      .eval {
+        Queue
+          .synchronousNoneTerminated[F2, Either[Token, Chunk[O]]]
+          .product(Ref[F2].of(F.unit -> false))
+      }
+      .flatMap {
+        case (q, currentTimeout) =>
+          def startTimeout: Stream[F2, Token] =
+            Stream.eval(F.delay(new Token)).evalTap { t =>
+              val timeout = timer.sleep(d) *> q.enqueue1(t.asLeft.some)
+
+              // Just using `supervise(timeout)` will ensure resource safety
+              // but finalisers accumulate and cause a leak, so we
+              // need to dispose of outstanding timeouts manually
+              timeout.start
+                .bracket(_ => F.unit) { fiber =>
```

Comment:
BTW, another technique is to use `Pull.acquireCancellable` or `Stream.bracketWithToken`, which lets us invoke a finalizer early *and* unregister it from the resource map.

Reply:
> which lets us invoke a finalizer early

ah! I actually didn't think of those combinators, but they sound promising. I might see how things look like with them.

Reply:
`bracketWithToken` is private[fs2] and fairly low level but a `bracketCancellable` might be useful to add to the API. Take a look at the implementation of `Pull.acquireCancellable` for an example of how to use `bracketWithToken` -- of particular note is the `release` call, which turns in to an `Algebra.Release(t)`, which is what prevents the finalizer leak.

Reply:
However, wouldn't that need a similar machinery to the code above, to  ensure that `release` is always called?

Reply:
Mm maybe not, it's not like `lease` right?
it's still bracketed, and _in addition to that_ it allows early finalisation?
if that's the case, it would be nice, but I'd rather we added a higher-level version to the public api before using it here

Reply:
Right. If `release(token)` is never called, the finalizer will still run at the normal place it would run with `bracket`. Note the similar in implementation: https://github.com/functional-streams-for-scala/fs2/blob/9a4074df5763341b82828c02d6eb9db5c5c260c9/core/shared/src/main/scala/fs2/Stream.scala#L2555-L2572

Reply:
Yeah, makes sense. I will still need the `Ref` to store the previous `(Token, R)`, but I could then deal with running those in the happy path with a simple flatMap, given that in case of problems (basically interruption) it will still be run

Reply:
Wishing I had known this before thinking so hard about synchronisation in the current code :P

Anyway, I'll add a `bracketCancelable` (or `earlyBracket` or whatever) in another PR, and then probably port this to use that instead, if the approach outlined in the comment right above makes sense to you

Reply:
Here's an example of a public API we could add:

```scala
  def bracketCaseCancellable[F[x] >: Pure[x], R](acquire: F[R])(
      release: (R, ExitCase[Throwable]) => F[Unit]): Stream[F, (Stream[F, INothing], R)] =
    bracketWithToken(acquire)(release).map {
      case (token, r) => (Stream.fromFreeC(Algebra.release(token, None)), r)
    } 
```